### PR TITLE
Fix AuthenticationResult ctor

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/AuthenticationResult.cs [hunk 1]
```diff
@@ -48,7 +48,7 @@ public partial class AuthenticationResult
             string idToken,
             IEnumerable<string> scopes,
             Guid correlationId,
-            AuthenticationResultMetadata authenticationResultMetadata,
+            AuthenticationResultMetadata authenticationResultMetadata = null,
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AuthenticationResult.cs:51]: Without a default value, this constitutes a breaking change which we failed to detect.

**Review by jmprieur — APPROVED** [review]
LGTM
Thanks Bogdan!

**Comment by henrik-me** [comment]
Please see latest comment from Scott Schaab

**Review by neha-bhargava — APPROVED** [review]
LGTM thanks!

**Review by henrik-me — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/AuthenticationResult.cs [hunk 2]
```diff
@@ -37,7 +38,7 @@ public partial class AuthenticationResult
         /// <param name="correlationId">The correlation id of the authentication request</param>
         /// <param name="tokenType">The token type, defaults to Bearer. Note: this property is experimental and may change in future versions of the library.</param>
         /// <param name="authenticationResultMetadata">Contains metadata related to the Authentication Result.</param>
-        public AuthenticationResult(
+        public AuthenticationResult( // for backwards compat with 4.17+
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/AuthenticationResult.cs:41]: >/ for backwards compat with 4.17+ [](start = 38, length = 33)

This should probably say compat with 4.17 to 4.20.1, and recommended ctor ?

File: src/client/Microsoft.Identity.Client/AuthenticationResult.cs [hunk 3]
```diff
@@ -65,6 +66,53 @@ public partial class AuthenticationResult
...
+          string uniqueId,
+          DateTimeOffset expiresOn,
+          DateTimeOffset extendedExpiresOn,
+          string tenantId,
+          IAccount account,
+          string idToken,
+          IEnumerable<string> scopes,
+          Guid correlationId,
+          AuthenticationResultMetadata authenticationResultMetadata,
+          string tokenType = "Bearer") :
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/AuthenticationResult.cs:98]: this seems to be the way it was in 4.17 to 4.20.1 ?

File: src/client/Microsoft.Identity.Client/AuthenticationResult.cs (hunk 2 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/AuthenticationResult.cs:41]: Correct.

File: src/client/Microsoft.Identity.Client/AuthenticationResult.cs (hunk 3 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/AuthenticationResult.cs:98]: Yes. To preserve backwards compat with those versions. I'll update the comment

**Review by schaabs — APPROVED** [review]
LGTM

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/AuthenticationResultTests.cs [hunk 4]
```diff
@@ -46,10 +46,69 @@ public void GetAuthorizationHeader()
...
+              DateTime.UtcNow,
+              "tid",
+              new Account("aid", "user", "env"),
+              "idt",
+              new[] { "scope" },
+              Guid.NewGuid(), 
+              "ProofOfBear");
+
+            Assert.IsNull(ar1.AuthenticationResultMetadata);
+            Assert.AreEqual("ProofOfBear", ar1.TokenType);
```
Comment (neha-bhargava) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/AuthenticationResultTests.cs:93]: I think it should be ar2.TokenType

File: src/client/Microsoft.Identity.Client/AuthenticationResult.cs (hunk 2 above)
Comment (henrik-me) [reply src/client/Microsoft.Identity.Client/AuthenticationResult.cs]: Still missing messaging around compat with 4.20.2+?

---
In reply to: [504958399](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2096#discussion_r504958399) [](ancestors = 504958399)

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/AuthenticationResultTests.cs [hunk 5]
```diff
@@ -46,10 +46,69 @@ public void GetAuthorizationHeader()
...
+                new AuthenticationResultMetadata(TokenSource.Cache));
 
             Assert.AreEqual("SomeTokenType at", ar.CreateAuthorizationHeader());
         }
+
+        [TestMethod]
+        [Description(
+            "In MSAL 4.17 we made a mistake and added AuthenticationResultMetadata with no default value before tokenType param. " +
+            "To fix this breaking change, we added 2 constructors - " +
+            "one for backwards compat with 4.17+ and one for 4.16 and below")]
```
Comment (henrik-me) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/AuthenticationResultTests.cs:59]: >4.16 and below [](start = 61, length = 14)

Additional comment: which is also the go forward recommended ctor, and way of initializing.
Similar meaning I would like to convey in the public API.

File: src/client/Microsoft.Identity.Client/AuthenticationResult.cs (hunk 2 above)
Comment (henrik-me) [reply src/client/Microsoft.Identity.Client/AuthenticationResult.cs]: Also, I think this should not be a code comment, but a remark. Recommended ctor or something similar?

---
In reply to: [505033029](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2096#discussion_r505033029) [](ancestors = 505033029,504958399)

File: src/client/Microsoft.Identity.Client/AuthenticationResult.cs [hunk 6]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.ComponentModel;
```
Comment (AlexGhiondea) [inline src/client/Microsoft.Identity.Client/AuthenticationResult.cs:6]: Do you know what is bringing this in?

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/AuthenticationResult.cs:6]: Yes, the attribute         `[EditorBrowsable(EditorBrowsableState.Never)]`
We use it elsewhere in the library as well.

**Comment by neha-bhargava** [comment]
Fixed in [MSAL 4.21.0 release](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/releases/tag/4.21.0)

**Pull request merged**